src/v/datalake/tests/datalake_avro_tests.cc

```diff
@@ -486,3 +491,708 @@ TEST(AvroSchema, TestLogicalTypes) {
+    // make sure iceberg holds a requested primitive value
+    if (!std::holds_alternative<iceberg::primitive_value>(current)) {
+        return AssertionFailure()
+               << fmt::format("Expected value {} to be a primitive", current);
+    }
+
+    auto& current_primitive = std::get<iceberg::primitive_value>(current);
+    if (!std::holds_alternative<IcebergValueT>(current_primitive)) {
+        return AssertionFailure() << fmt::format(
+                 "Expected value {} to be of type {}",
+                 current,
+                 IcebergValueT{});
+    }
+    return AssertionSuccess();
+}
+
+template<typename IcebergT, typename Matcher>
+AssertionResult primitive_equal(
+  const avro::GenericDatum& expected,
+  const iceberg::value& current,
+  Matcher matcher_func) {
+    // make sure iceberg holds a requested primitive value
+    if (!std::holds_alternative<iceberg::primitive_value>(current)) {
+        return AssertionFailure()
+               << fmt::format("Expected value {} to be a primitive", current);
+    }
+
+    auto& current_primitive = std::get<iceberg::primitive_value>(current);
+    if (!std::holds_alternative<IcebergT>(current_primitive)) {
+        return AssertionFailure() << fmt::format(
+                 "Expected value {} to be of type {}", current, IcebergT{});
+    }
+
+    const auto& current_value = std::get<IcebergT>(current_primitive);
+
+    return matcher_func(current_value, expected);
+}
+
+AssertionResult value_matches(
+  const avro::GenericDatum& datum, const iceberg::value& value, bool is_union);
+
+AssertionResult value_matches(
+  const avro::GenericDatum& datum,
+  const std::optional<iceberg::value>& optional_val,
+  bool is_union) {
+    if (optional_val) {
+        return value_matches(datum, optional_val.value(), is_union);
+    }
+    if (datum.type() != avro::AVRO_NULL) {
+        return AssertionFailure() << "Empty optional must be represented as "
+                                     "null in AVRO";
+    }
+    return AssertionSuccess();
+}
+
+// Helper macro to minimize boilerplate code
+#define ASSERT_VALUES_EQUAL(current, expected)                                 \
+    if (current != expected) {                                                 \
+        return AssertionFailure() << fmt::format(                              \
+                 "Expected value {} to be equal to {}", current, expected);    \
+    }                                                                          \
+    return AssertionSuccess();
+
+// convert iceberg decimal to vector of bytes, big endian
+std::vector<uint8_t> decimal_to_vector(absl::int128 decimal) {
+    auto high_half = absl::Int128High64(decimal);
+    auto low_half = absl::Int128Low64(decimal);
+
+    std::vector<uint8_t> decimal_bytes(16, 0);
+    for (int i = 0; i < 8; i++) {
+        decimal_bytes[i] = (high_half >> ((7 - i) * 8)) & 0xFF;
+        decimal_bytes[i + 8] = (low_half >> ((7 - i) * 8)) & 0xFF;
+    }
```

Comment:
I feel like we've repeated this conversion in a number of places... Should we have some helpers in utils or something that goes absl::int128<->std::array<uint8_t, 16>, then you can copy the array from the stack to an iobuf or vector if you want?